tests/study_tests/test_optimize.py

```diff
@@ -0,0 +1,143 @@
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Optional
+
+import pytest
+
+from optuna import _optimize
+from optuna import create_study
+from optuna import Study
+from optuna import Trial
+from optuna import TrialPruned
+from optuna.exceptions import TrialPruned as TrialPruned_in_exceptions
+from optuna.structs import TrialPruned as TrialPruned_in_structs
+from optuna.testing.storage import StorageSupplier
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+STORAGE_MODES = [
+    "inmemory",
+    "sqlite",
+    "cache",
+    "redis",
+]
+
+
+def func(trial: Trial, x_max: float = 1.0) -> float:
+
+    x = trial.suggest_uniform("x", -x_max, x_max)
+    y = trial.suggest_loguniform("y", 20, 30)
+    z = trial.suggest_categorical("z", (-1.0, 1.0))
+    assert isinstance(z, float)
+    return (x - 2) ** 2 + (y - 25) ** 2 + z
+
+
+def check_params(params: Dict[str, Any]) -> None:
+
+    assert sorted(params.keys()) == ["x", "y", "z"]
+
+
+def check_value(value: Optional[float]) -> None:
+
+    assert isinstance(value, float)
+    assert -1.0 <= value <= 12.0 ** 2 + 5.0 ** 2 + 1.0
+
+
+def check_frozen_trial(frozen_trial: FrozenTrial) -> None:
+
+    if frozen_trial.state == TrialState.COMPLETE:
+        check_params(frozen_trial.params)
+        check_value(frozen_trial.value)
+
+
+def check_study(study: Study) -> None:
+
+    for trial in study.trials:
+        check_frozen_trial(trial)
+
+    assert not study._is_multi_objective()
+
+    complete_trials = [t for t in study.trials if t.state == TrialState.COMPLETE]
+    if len(complete_trials) == 0:
+        with pytest.raises(ValueError):
+            study.best_params
+        with pytest.raises(ValueError):
+            study.best_value
+        with pytest.raises(ValueError):
+            study.best_trial
+    else:
+        check_params(study.best_params)
+        check_value(study.best_value)
+        check_frozen_trial(study.best_trial)
```

Comment:
These objective and check functions are copied from `test_study.py`. It is redundant. I think we can move these to `optuna/testing`. Do you have any opinions?

Reply:
Your suggestion makes sense. I'm in favor of moving.

Reply:
These tests are mostly overlapping with `test_study.py` so we can also relax the tests here. These helpers are not very general in my opinion with hard coded values.